### PR TITLE
Add SU2-FEniCS Perpendicular Flap FSI Tutorial

### PR DESCRIPTION
**Pull request by IshaanDesai**
For the perpendicular flap FSI tutorial we have the combinations:
`OpenFOAM-FEniCS`
`OpenFOAM-CalculiX`
`SU2-CalculiX` 
To complete all combinations and show the true black-box use of preCICE and the official adapters, this PR adds the `SU2-FEniCS` tutorial case.

### CODE REVIEW

**Comment by IshaanDesai** [comment]
This mesh used for Fluid participant (SU2) is constructed in GMSH and has the structure:
SU2_mesh

**Comment by IshaanDesai** [comment]
The folder [initial-run](https://github.com/precice/tutorials/tree/14a5ad8adefe28c23bb5f8f40df1bf861a401f3e/FSI/flap_perp/SU2-FEniCS/initial_run) gives the setup used to run a SU2 simulation with the flap as a rigid body. The resulting flow profile is used as a reference to run the coupled FSI case.

File: FSI/flap_perp/SU2-FEniCS/FEniCS.log [hunk 1]
```diff
@@ -0,0 +1,5 @@
+---[precice] [0m This is preCICE version 2.0.2
```
Comment (MakisH) [inline FSI/flap_perp/SU2-FEniCS/FEniCS.log:1]: This file should probably not be here.

File: FSI/flap_perp/SU2-FEniCS/Output/plotDisplacement.sh [hunk 2]
```diff
@@ -0,0 +1,8 @@
+#! /bin/bash
+gnuplot -p << EOF
+set grid
+set title 'Displacement of the Flap Tip'
+set xlabel 'Time [s]'
+set ylabel 'X-Displacement [m]'
+plot "point1.watchpoint.txt" using 1:5 smooth cumulative title ""
```
Comment (MakisH) [inline FSI/flap_perp/SU2-FEniCS/Output/plotDisplacement.sh:7]: This name has changed in recent preCICE versions:
```suggestion
plot "precice-Calculix-watchpoint-point1.log using 1:5 smooth cumulative title ""
```
we need to update it elsewhere as well. See changes proposed in https://github.com/precice/tutorials/pull/100.

File: FSI/flap_perp/SU2-FEniCS/Su2.log [hunk 3]
```diff
@@ -0,0 +1,198 @@
+
```
Comment (MakisH) [inline FSI/flap_perp/SU2-FEniCS/Su2.log:1]: This file should probably not be here.

File: FSI/flap_perp/SU2-FEniCS/plotDisplacement.sh [hunk 4]
```diff
@@ -0,0 +1,8 @@
+#! /bin/bash
```
Comment (MakisH) [inline FSI/flap_perp/SU2-FEniCS/plotDisplacement.sh:1]: Why do we have this file two times? I guess we should only keep this one (with the updates mentioned in the one under `Output/`).

**Review by MakisH — CHANGES_REQUESTED** [review]
Thank you very much for adding this! :+1:

There have been quite some changes (about to merge) in the run scripts and configuration, see https://github.com/precice/tutorials/pull/100. Please wait for this to be merged and then port the same changes here.

It would be good to also remove a few log files that slipped through.

**Comment by IshaanDesai** [comment]
Thanks for the feedback @MakisH ! Will look out for #100 before merging this one. I am still working on this PR because currently the simulation diverges

**Comment by MakisH** [comment]
@IshaanDesai the SU2-CalculiX also diverges (at least under some systems), so this is not necessarily FEniCS-specific.

**Comment by IshaanDesai** [comment]
While tuning the SU2-FEniCS case and analysing results, it was found that the su2-adapter works with DisplacementDeltas and the fenics-adapters works with Displacements. To get a working tutorial, the fenics-adapter would need to be modified to handle DisplacementDeltas

**Comment by MakisH** [comment]
There is also a related issue on adding Displacements to SU2: https://github.com/precice/su2-adapter/issues/17

**Comment by IshaanDesai** [comment]
Thanks @MakisH for pointing that out! After discussions with @uekerman and @BenjaminRueth it is decided to currently resolve this tutorial via handling `DisplacementDeltas` in the `fenics-adapter`

**Comment by IshaanDesai** [comment]
After further discussions it was concluded that handling the absolute Displacement values or delta Displacement values is something that a FEniCS user can do in the FEniCS case file. Due to the unique way in which FEniCS is set up and the adapter is constructed it was not found unworthy to implement the handling of `DisplacementDeltas` in the adapter.

**Comment by IshaanDesai** [comment]
Tip displacement plot:
tipDispl_SU2_FEniCS
The non-smooth behaviour is because the structure solver in FEniCS is a linear solver.

**Review by BenjaminRueth — COMMENTED** [review]
I was not able to run the tutorial, since FEniCS suddenly stopped working on my system and I could not run the tutorial. The only thing I can contribute for now is a minor error in documentation that I spotted. I now first have to fix my FEniCS installation.

**Review by BenjaminRueth — COMMENTED** [review]
After building FEniCS from source I finally got it working again.

Running the tutorial worked for me well using `./runSolid` and `./runFluid`. I was also able to visualize the results.

After 500 timesteps the Solid solver crashed with the following error:

```
---[precice]  it 1 of 50 | dt# 501 of 600 | t 5 | dt 0.01 | max dt 0.01 | ongoing yes | dt complete yes | write-iteration-checkpoint | 

Found a double-boundary point at (0.05, 0.0).
Found a double-boundary point at (-0.05, 0.0).
Found a double-boundary point at (0.05, 0.0).
Found a double-boundary point at (-0.05, 0.0).
Found a double-boundary point at (0.05, 0.0).
Found a double-boundary point at (-0.05, 0.0).
Found a double-boundary point at (0.05, 0.0).
Found a double-boundary point at (-0.05, 0.0).
Found a double-boundary point at (0.05, 0.0).
Found a double-boundary point at (-0.05, 0.0).
Found a double-boundary point at (0.05, 0.0).
Found a double-boundary point at (-0.05, 0.0).
Found a double-boundary point at (0.05, 0.0).
Found a double-boundary point at (-0.05, 0.0).
Found a double-boundary point at (0.05, 0.0).
Found a double-boundary point at (-0.05, 0.0).
Found a double-boundary point at (0.05, 0.0).
Found a double-boundary point at (-0.05, 0.0).
Found a double-boundary point at (0.05, 0.0).
Found a double-boundary point at (-0.05, 0.0).
Found a double-boundary point at (0.05, 0.0).
Found a double-boundary point at (-0.05, 0.0).
Found a double-boundary point at (0.05, 0.0).
Found a double-boundary point at (-0.05, 0.0).
Found a double-boundary point at (0.05, 0.0).
Found a double-boundary point at (-0.05, 0.0).
Found a double-boundary point at (0.05, 0.0).
Found a double-boundary point at (-0.05, 0.0).
Found a double-boundary point at (0.05, 0.0).
Couldn't map 'f_1182' to a float, returning ufl object without evaluation.
Couldn't map 'f_1182' to a float, returning ufl object without evaluation.
Couldn't map 'f_1182' to a float, returning ufl object without evaluation.
Couldn't map 'f_1182' to a float, returning ufl object without evaluation.
Couldn't map 'f_1182' to a float, returning ufl object without evaluation.
Couldn't map 'f_1182' to a float, returning ufl object without evaluation.
Couldn't map 'f_1182' to a float, returning ufl object without evaluation.
Couldn't map 'f_1182' to a float, returning ufl object without evaluation.
Couldn't map 'f_1182' to a float, returning ufl object without evaluation.
Couldn't map 'f_1182' to a float, returning ufl object without evaluation.
Couldn't map 'f_1182' to a float, returning ufl object without evaluation.
Couldn't map 'f_1182' to a float, returning ufl object without evaluation.
Couldn't map 'f_1182' to a float, returning ufl object without evaluation.
Couldn't map 'f_1182' to a float, returning ufl object without evaluation.
terminate called after throwing an instance of 'pybind11::error_already_set'
  what():  RecursionError: maximum recursion depth exceeded while calling a Python object

At:
  /home/benjamin/.local/lib/python3.6/site-packages/ufl/core/operator.py(61): <genexpr>

[benjamin-ThinkPad-X1-Yoga-2nd:11711] *** Process received signal ***
[benjamin-ThinkPad-X1-Yoga-2nd:11711] Signal: Aborted (6)
[benjamin-ThinkPad-X1-Yoga-2nd:11711] Signal code:  (-6)
[benjamin-ThinkPad-X1-Yoga-2nd:11711] [ 0] /lib/x86_64-linux-gnu/libc.so.6(+0x3efd0)[0x7f0c3a75cfd0]
[benjamin-ThinkPad-X1-Yoga-2nd:11711] [ 1] /lib/x86_64-linux-gnu/libc.so.6(gsignal+0xc7)[0x7f0c3a75cf47]
[benjamin-ThinkPad-X1-Yoga-2nd:11711] [ 2] /lib/x86_64-linux-gnu/libc.so.6(abort+0x141)[0x7f0c3a75e8b1]
[benjamin-ThinkPad-X1-Yoga-2nd:11711] [ 3] /usr/lib/x86_64-linux-gnu/libstdc++.so.6(+0x8c957)[0x7f0c3594e957]
[benjamin-ThinkPad-X1-Yoga-2nd:11711] [ 4] /usr/lib/x86_64-linux-gnu/libstdc++.so.6(+0x92ae6)[0x7f0c35954ae6]
[benjamin-ThinkPad-X1-Yoga-2nd:11711] [ 5] /usr/lib/x86_64-linux-gnu/libstdc++.so.6(+0x92b21)[0x7f0c35954b21]
[benjamin-ThinkPad-X1-Yoga-2nd:11711] [ 6] /usr/lib/x86_64-linux-gnu/libstdc++.so.6(+0x92d54)[0x7f0c35954d54]
[benjamin-ThinkPad-X1-Yoga-2nd:11711] [ 7] /home/benjamin/.local/lib/python3.6/site-packages/dolfin/cpp.cpython-36m-x86_64-linux-gnu.so(+0x292fa8)[0x7f0c38da6fa8]
[benjamin-ThinkPad-X1-Yoga-2nd:11711] [ 8] python3(_PyCFunction_FastCallDict+0x35c)[0x566ddc]
[benjamin-ThinkPad-X1-Yoga-2nd:11711] [ 9] python3[0x50a783]
[benjamin-ThinkPad-X1-Yoga-2nd:11711] [10] python3(_PyEval_EvalFrameDefault+0x444)[0x50c1f4]
[benjamin-ThinkPad-X1-Yoga-2nd:11711] [11] python3(_PyFunction_FastCallDict+0xf5)[0x509015]
[benjamin-ThinkPad-X1-Yoga-2nd:11711] [12] python3[0x594b01]
[benjamin-ThinkPad-X1-Yoga-2nd:11711] [13] python3(PyObject_Call+0x3e)[0x59fe1e]
[benjamin-ThinkPad-X1-Yoga-2nd:11711] [14] python3[0x4b1eea]
[benjamin-ThinkPad-X1-Yoga-2nd:11711] [15] python3[0x488370]
[benjamin-ThinkPad-X1-Yoga-2nd:11711] [16] python3[0x59cb00]
[benjamin-ThinkPad-X1-Yoga-2nd:11711] [17] python3(_PyEval_EvalFrameDefault+0x28bc)[0x50e66c]
[benjamin-ThinkPad-X1-Yoga-2nd:11711] [18] python3[0x507f24]
[benjamin-ThinkPad-X1-Yoga-2nd:11711] [19] python3(PyEval_EvalCode+0x23)[0x50b053]
[benjamin-ThinkPad-X1-Yoga-2nd:11711] [20] python3[0x634dd2]
[benjamin-ThinkPad-X1-Yoga-2nd:11711] [21] python3(PyRun_FileExFlags+0x97)[0x634e87]
[benjamin-ThinkPad-X1-Yoga-2nd:11711] [22] python3(PyRun_SimpleFileExFlags+0x17f)[0x63863f]
[benjamin-ThinkPad-X1-Yoga-2nd:11711] [23] python3(Py_Main+0x591)[0x6391e1]
[benjamin-ThinkPad-X1-Yoga-2nd:11711] [24] python3(main+0xe0)[0x4b0dc0]
[benjamin-ThinkPad-X1-Yoga-2nd:11711] [25] /lib/x86_64-linux-gnu/libc.so.6(__libc_start_main+0xe7)[0x7f0c3a73fb97]
[benjamin-ThinkPad-X1-Yoga-2nd:11711] [26] python3(_start+0x2a)[0x5b26fa]
[benjamin-ThinkPad-X1-Yoga-2nd:11711] *** End of error message ***
```
Any idea what this means? Note: The error might also be caused by my broken system. I see the same error (also after 500 timesteps) for the OpenFOAM-FEniCS tutorial.

I cannot say much about the SU2 part of the new tutorial. The FEniCS part looks good to me.

One minor thing I find distracting is the message 

```
Found a double-boundary point at (-0.05, 0.0).
Found a double-boundary point at (0.05, 0.0).
```

that occasionally pops up. But I think we should discuss this issue elsewhere :arrow_right:  https://github.com/precice/fenics-adapter/issues/90

**Review by uekerman — CHANGES_REQUESTED** [review]
Thanks for adding this :+1:
Some remarks and small tweaks:

* I get the same crash as @BenjaminRueth. 

* To be consistent to how we organize tutorials currently, the case should better go into `flap_perp_2D`. Soonish we want to re-organize such things anyway (#28), but for now.

* Could you please [format](https://github.com/precice/precice/blob/develop/tools/formatting/config-formatter) the preCICE config and remove all commented out parts there.

* We should carry the case to a valid regime of the linear structural model. So, smaller displacements. For example by increasing the E module.

**Comment by IshaanDesai** [comment]
Increasing the modulus of elasticity (E) for the structure from 4E5 to 1E6 leads to a smoother tip displacement plot:
tipDispl_plot

**Comment by MakisH** [comment]
> The 2D mesh used for Fluid participant (SU2) is constructed in GMSH and has the structure:
> 
> Screenshot from 2020-09-01 16-23-10

Could you please describe here how you generated this mesh? Could be useful in the future.

**Review by MakisH — APPROVED** [review]
Excellent! All the configuration files look good, see a few minor comments. :+1:

I cannot run the case at all at the moment (system-specific issue with mpi4py), but since it works for you, go ahead.

Remember to integrate this case into the wiki.

I quickly skimmed through the FEniCS solver code, but I cannot see where the recursion could be originating from. The stack trace also does not help much. Maybe this is originating from the Python bindings?

The SU2-CalculiX case is also unreliable right now, so a convergence issue would not be necessarily a FEniCS issue.

**Comment by MakisH** [comment]
It would be good to also add this case afterwards to our system tests. We can wait for https://github.com/precice/tutorials/issues/109, or we can simply test only a few seconds.

**Review by uekerman — APPROVED** [review]

**Review by BenjaminRueth — APPROVED** [review]

File: FSI/flap_perp_2D/SU2-FEniCS/Solid/perp-flap.py [hunk 5]
```diff
@@ -0,0 +1,249 @@
...
+
+    # Either revert to old step if timestep has not converged or move to next timestep
+    if precice.is_action_required(precice.action_read_iteration_checkpoint()):  # roll back to checkpoint
+        u_cp, t_cp, n_cp = precice.retrieve_checkpoint()
+        u_n.assign(u_cp)
+        t = t_cp
+        n = n_cp
+    else:
+        u_n.assign(u_np1)
+        t += dt
```
Comment (BenjaminRueth) [inline FSI/flap_perp_2D/SU2-FEniCS/Solid/perp-flap.py:230]:
```suggestion
        t += float(dt)
```
See https://github.com/precice/tutorials/pull/115

**Review by BenjaminRueth — CHANGES_REQUESTED** [review]

**Pull request merged**